File: src/clase-06/typescript/observer/test_observer.ts
import { Stock, EmailAlert, Dashboard } from './observer';

function captureConsole(fn: () => void): string[] {
  const logs: string[] = [];
  const orig = console.log;
  console.log = (...args: any[]) => { logs.push(args.join(' ')); };
  try { fn(); } finally { console.log = orig; }
  return logs;
}

const logs = captureConsole(() => {
  const s = new Stock();
  const e = new EmailAlert();
  const d = new Dashboard();
  s.subscribe(e); s.subscribe(d);
  s.setPrice(10.5);
});

if (!logs.some(x => x.includes('Nuevo precio: 10.5'))) throw new Error('EmailAlert no recibió notificación');
if (!logs.some(x => x.includes('Precio actualizado: 10.5'))) throw new Error('Dashboard no recibió notificación');

console.log('test_observer OK');

File: src/clase-06/python/observer/test_observer.py
import builtins
from python.observer.observer import Stock, EmailAlert, Dashboard

def test_observer_notify(monkeypatch, capsys):
    s = Stock()
    e = EmailAlert()
    d = Dashboard()
    s.subscribe(e)
    s.subscribe(d)

    s.set_price(10.5)
    out = capsys.readouterr().out
    assert "[Email] Nuevo precio: 10.5" in out
    assert "[Dashboard] Precio actualizado: 10.5" in out

def test_unsubscribe(monkeypatch, capsys):
    s = Stock()
    e = EmailAlert()
    s.subscribe(e)
    s.unsubscribe(e)

    s.set_price(99.9)
    out = capsys.readouterr().out
    assert out.strip() == ""
